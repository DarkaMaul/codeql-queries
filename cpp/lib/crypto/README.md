# Overview

This directory contains CodeQL base types targeted at C and C++
cryptographic implementations. The code is organized as follows:

  - `common`: Common, abstract base types used for general queries. These are collected
    in the QL library `common.qll`.
  - `libraries`: Library specific implementations of abstract base types
    defined under `common`. The library currently contains concrete
    implementations for parts of the `OpenSSL` and `mbedtls` libraries. Each
    library exposes a `<LIBRARY NAME>.qll` library which can be
    imported to target that specific library. When targeting all libraries,
    import `libraries.qll`.

A goal of this project is to attempt to define reusable, abstract base types
with enough expressive power to describe many common cryptographic
vulnerabilities across many different implementations. This is still very much
a work in progress which means that the structure and definitions of these
types will most likely change over time.
