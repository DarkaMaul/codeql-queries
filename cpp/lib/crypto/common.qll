import common.InitUpdateFinalPattern
import common.SymmetricCipherContext

import common.HashFunction
import common.CustomAllocator
import common.CsprngInitializer
import common.StrongPasswordSink
import common.StrongRandomnessSink
import common.WeakRandomnessSource
import common.StaticCryptoVariable
import common.KeyDerivationFunction
