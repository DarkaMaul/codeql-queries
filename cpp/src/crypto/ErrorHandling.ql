/**
 * @name RAND_bytes and BN_rand error handling
 * @id tob/cpp/error-handling
 * @description Checks if care is being taken to perform proper error handling
 * @kind problem
 * @tags correctness crypto
 * @problem.severity warning
 * @precision high
 * @group cryptography
 */

import cpp
import semmle.code.cpp.dataflow.new.DataFlow

predicate isChecked(Expr value) {
    // The return value flows into the condition of an if-statement.
    exists (IfStmt is |
        DataFlow::localFlow(
            DataFlow::exprNode(value),
            DataFlow::exprNode(is.getCondition().getAChild*())
        )
    ) or
    // The return value flows into the condition of a while-statement.
    exists (WhileStmt ws |
        DataFlow::localFlow(
            DataFlow::exprNode(value),
            DataFlow::exprNode(ws.getCondition().getAChild*())
        )
    ) or
    // The return value flows into the condition of a switch-statement.
    exists (SwitchStmt ss |
        DataFlow::localFlow(
            DataFlow::exprNode(value),
            DataFlow::exprNode(ss.getExpr().getAChild*())
        )
    )
}

// TODO: This should use an abstract type for functions that return status codes.
predicate isReturnValue(Expr value){
    value.(FunctionCall).getTarget().getName() in ["BN_rand", "RAND_bytes"]
}


from FunctionCall call
where isReturnValue(call)
    and not isChecked(call)
select call.getLocation(), "The function fails to check the return value of '" + call.getTarget().getName() + "'"
