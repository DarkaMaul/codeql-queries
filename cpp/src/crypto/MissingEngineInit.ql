/**
 * @name Missing engine initialization
 * @id tob/cpp/missing-engine-init
 * @description Finds created OpenSSL engines that may not be properly initialized
 * @kind problem
 * @tags correctness crypto
 * @problem.severity warning
 * @precision medium
 * @group cryptography
 */

import cpp
import semmle.code.cpp.dataflow.new.DataFlow

// REFERENCE: https://www.openssl.org/docs/man1.1.1/man3/ENGINE_init.html
//
// ENGINE_new, ENGINE_by_id, etc, all create a new engine object.
class CreateEngine extends FunctionCall {
  CreateEngine() {
    getTarget().getName() in ["ENGINE_new", "ENGINE_by_id", "ENGINE_get_next", "ENGINE_get_prev"]
  }
}

// ENGINE_init is called to initialize the engine.
class ENGINE_Init extends FunctionCall {
  ENGINE_Init() { getTarget().getName() = "ENGINE_init" }

  Expr getEngine() { result = this.getArgument(0) }
}

// ENGINE_set_default should usually be called to enable the provided algorithms as defaults.
class ENGINE_set_default extends FunctionCall {
  ENGINE_set_default() { getTarget().getName() = "ENGINE_set_default" }

  Expr getEngine() { result = this.getArgument(0) }
}

predicate flowsToEngineInit(CreateEngine create) {
  exists(ENGINE_Init init |
    DataFlow::localFlow(DataFlow::exprNode(create), DataFlow::exprNode(init.getEngine()))
  )
}

predicate flowsToEngineSet(CreateEngine create) {
  exists(ENGINE_set_default set |
    DataFlow::localFlow(DataFlow::exprNode(create), DataFlow::exprNode(set.getEngine()))
  )
}

from CreateEngine createCall
where not flowsToEngineInit(createCall) or not flowsToEngineSet(createCall)
select createCall.getLocation(), "Engine may not be properly initialized"
