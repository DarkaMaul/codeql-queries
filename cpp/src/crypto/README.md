# Overview

This directory contains CodeQL queries targeted at C and C++ cryptographic implementations.

The CodeQL queries are based on the abstract base types defined in `trailofbits/cpp-all` library.
The intention is that they should be easy to extend to other cryptographic libraries written in either C or C++.
