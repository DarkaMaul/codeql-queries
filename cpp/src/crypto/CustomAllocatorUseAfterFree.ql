/**
 * @name Memory use after free related to custom allocator
 * @id tob/cpp/custom-allocator-use-after-free
 * @description Finds use-after-frees related to custom allocators like `BN_new`
 * @kind problem
 * @tags correctness crypto
 * @problem.severity warning
 * @precision medium
 * @group cryptography
 */

import cpp
import trailofbits.crypto.libraries

string toLine(ControlFlowNode node) { result = node.getLocation().getStartLine().toString() }

from
  CustomAllocatorUseAfterFree uaf, SemanticStackVariable var, ControlFlowNode source,
  ControlFlowNode sink
where uaf.reaches(source, var, sink)
select source.getLocation(),
  "The variable " + var.toString() + " is deallocated on line " + toLine(source) +
    " and then subsequently used on line " + toLine(sink)
