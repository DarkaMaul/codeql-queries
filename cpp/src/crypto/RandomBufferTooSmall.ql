/**
 * @name Random buffer too small
 * @id tob/cpp/random-buffer-too-small
 * @description Finds buffer overflows in calls to `RAND_bytes`
 * @kind problem
 * @tags security crypto
 * @problem.severity warning
 * @precision high
 * @group cryptography
 */

import cpp
import crypto.libraries

predicate requestedRandomnessFitsInBuffer(RAND_bytes call) { call.getRequestedBytes() <= call.getBufferSize() }

predicate bufferIsPointer(RAND_bytes call) {
    call.getBuffer().getType().getPointerIndirectionLevel() > 0
}

from RAND_bytes randBytesCall
where not bufferIsPointer(randBytesCall) and not requestedRandomnessFitsInBuffer(randBytesCall)
select randBytesCall.getLocation(), "Buffer size (" + randBytesCall.getBufferSize() + ") is smaller than the number of requested bytes (" + randBytesCall.getRequestedBytes() + ") in call to RAND_bytes."
