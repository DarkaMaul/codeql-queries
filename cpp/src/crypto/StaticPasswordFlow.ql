/**
 * @name Crypto variable initialized using static password
 * @id tob/cpp/static-password-flow
 * @description Finds crypto variables initialized using static passwords
 * @kind problem
 * @tags security crypto
 * @problem.severity error
 * @precision high
 * @group cryptography
 */

import cpp
import trailofbits.crypto.libraries
import semmle.code.cpp.dataflow.new.DataFlow

from
  DataFlow::Node source, DataFlow::Node sink,
  StrongPasswordSink random
where
  StaticPasswordFlow::flow(source, sink) and
  random = sink.asExpr()
select sink.getLocation(),
  random.getDescription() + " is initialized using static data from " + source.toString()
