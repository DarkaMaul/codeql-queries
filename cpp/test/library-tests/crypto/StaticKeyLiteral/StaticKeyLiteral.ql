import cpp
import crypto.libraries


from
  StaticKeyLiteral key
select
  key
