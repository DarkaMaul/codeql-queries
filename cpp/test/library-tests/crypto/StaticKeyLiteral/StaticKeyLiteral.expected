| test.c:12:25:15:1 | {...} |
| test.c:16:24:18:1 | {...} |
| test.c:92:23:92:28 | {...} |
| test.c:93:24:93:29 | {...} |
